Simplify WooFoundation into a Swift package

mokagio opened the pull request:
## Description
What it says in the title... 

Notice that the makeup changed from the frameworks. We had WooFoundation and WooFoundationWatchOS, with WooFoundation including all the files in WooFoundationWatchOS.

Now, we have WooFoundationCore which has only the files required in watchOS and is necessarily UIKit free. WooFoundation depends WooFoundationCore. A few additional comments in the diff.

As expected, this also makes the fix implemented in https://github.com/woocommerce/woocommerce-ios/pull/15673 redundant, cc @staskus . I'm pointing this out not to discredit the fix but just to highlight how odd that linking behavior was. I am glad we implemented the fix, even if it stayed up just a few days. Given we couldn't know how long it would have taken for this PR to be ready and land, it was better to address the issue ASAP.

## Steps to reproduce & Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
See green CI and run on device.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

dangermattic commented:
<!--
  0 Errors
  3 Warnings: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Temporarily reorganize WooFoundation sources in folders before SPM

mokagio committed:
Simplify WooFoundation into a Swift package - With FIXMEs

mokagio committed:
Reorder products and targets list alphabetical order

mokagio committed:
Rename WooFoundationLite to WooFoundationCore

mokagio committed:
Address FIXME in WooFoundation left during migration

In migrating from an iOS vs watchOS framework architecture to a
cross-platform package one, we had to change how the code deals with
accessing the `Bundle`. In particular, the WooFoundation `UIColor` code
is no longer in the same `Bundle` where the palette is.

I chose to expose the palette's `Bundle` via a property on
`ColorStudio`.  This makes the `Bundle` instance accessible in
WooFoundation so that WooFoundationCore can remain UIKit free and run on
watchOS.

mokagio commented:
🤔  the unit tests crash in CI but the fix works on my end...

mokagio commented:
I stand corrected... The test pass when running from the standalone scheme of the package
<img width="1461" alt="image" src="https://github.com/user-attachments/assets/fc9119b4-a60f-477e-817e-172269b48d4b" />

but crash when running from the WooCommerce scheme test action

![image](https://github.com/user-attachments/assets/e62fd6d9-f896-4de6-b2c3-34bb9a270815)

---

Update: 25a4d5bb4fd12062c8aafd44b3ee864818e39318 addresses the issue.

mokagio committed:
Add workaround for XCtest to WooFoundationCore `Bundle`

mokagio commented on Modules/Package.swift:
```diff
@@ -339,29 +361,12 @@ enum XcodeSupport {
                     .product(name: "Sentry", package: "sentry-cocoa"),
                 ]
             ),
-            .xcodeTarget(
-                XcodeTargetNames.wooFoundation,
-                dependencies: [
-                    "Codegen",
-                    .product(name: "CocoaLumberjackSwift", package: "CocoaLumberjack")
-                ]
-            ),
-            .xcodeTarget(
-                XcodeTargetNames.wooFoundationTests,
-                dependencies: ["TestKit", XcodeTargetNames.wooFoundation.asDependency]
-            ),
-            .xcodeTarget(
-                XcodeTargetNames.wooFoundationWatchOS,
-                dependencies: [
-                    "Codegen",
-                    .product(name: "CocoaLumberjackSwift", package: "CocoaLumberjack")
-                ]
-            ),
```
As promised in #15654 , these `XcodeTarget_...` are beginning to disappear in favor of packages. The only ones that will remain are those for the apps and extensions targets.

mokagio commented on Modules/Sources/WooFoundation/Colors/UIColor+ColorStudio.swift:
```diff
@@ -9,7 +9,7 @@ public extension UIColor {
     class func withColorStudio(_ colorStudio: ColorStudio) -> UIColor {
         let assetName = colorStudio.assetName()
-        let color = UIColor(named: assetName, in: Bundle(for: WooFoundationBundleClass.self), compatibleWith: nil)
+        let color = UIColor(named: assetName, in: ColorStudio.bundle, compatibleWith: nil)
```
As part of the restructure, the color palette went into WooFoundationCore. In Swift packages, we can't access bundles via `Bundle(for:)`, rather, packages with resources generate a `Bundle.module` helper. However, this is only accessible via that API from within the package. In order to load the color for a `UIColor` in WooFoundation (because WooFoundationCore cannot `import UIKit`), I added this `.bundle` property on `ColorStudio`.

See the implementation below for more details.

mokagio commented on WooCommerce/WooCommerceTests/UnitTests.xctestplan:
```diff
@@ -32,8 +32,8 @@
-        "identifier" : "WordPressSharedObjCTests",
-        "name" : "WordPressSharedObjCTests"
+        "identifier" : "WooFoundationTests",
+        "name" : "WooFoundationTests"
```
I am yet to find a way to keep Xcode from shuffling the `xctestplan` every time a target gets added. Sorry for the noise.

staskus reviewed (approved):
🚀 Thanks, works well! Tested on a device.

> As expected, this also makes the fix implemented in https://github.com/woocommerce/woocommerce-ios/pull/15673 redundant, cc @staskus . I'm pointing this out not to discredit the fix but just to highlight how odd that linking behavior was.

Well... We couldn't find a good explanation why we required this explicit linking so it's good to see that it's no longer necessary 👍
